iothub/service/src/Amqp/AmqpClientHelper.cs

```diff
@@ -33,109 +31,170 @@ public static Exception ToIotHubClientContract(Exception exception)
+            {
+                retException = new MessageTooLargeException(message);
+            }
+            else if (error.Condition.Equals(AmqpErrorCode.ResourceLimitExceeded))
+            {
+                retException = new DeviceMaximumQueueDepthExceededException(message);
+            }
+            else if (error.Condition.Equals(IotHubAmqpErrorCode.DeviceAlreadyExists))
+            {
+                retException = new DeviceAlreadyExistsException(message, (Exception)null);
+            }
+            else if (error.Condition.Equals(IotHubAmqpErrorCode.DeviceContainerThrottled))
+            {
+                retException = new IotHubThrottledException(message, (Exception)null);
+            }
+            else if (error.Condition.Equals(IotHubAmqpErrorCode.QuotaExceeded))
+            {
+                retException = new QuotaExceededException(message, (Exception)null);
+            }
+            else if (error.Condition.Equals(IotHubAmqpErrorCode.PreconditionFailed))
+            {
+                retException = new PreconditionFailedException(message, (Exception)null);
+            }
+            else if (error.Condition.Equals(IotHubAmqpErrorCode.IotHubSuspended))
+            {
+                retException = new IotHubSuspendedException(message);
+            }
+            else
+            {
+                retException = new IotHubException(message);
             }
-        }
 
-        public static void ValidateContentType(AmqpMessage amqpMessage, string expectedContentType)
-        {
-            string contentType = amqpMessage.Properties.ContentType.ToString();
-            if (!string.Equals(contentType, expectedContentType, StringComparison.OrdinalIgnoreCase))
+            if (trackingId != null
+                && retException is IotHubException exHub)
             {
-                throw new InvalidOperationException("Unsupported content type: {0}".FormatInvariant(contentType));
+                IotHubException iotHubException = exHub;
+                iotHubException.TrackingId = trackingId;
+                // This is created but not assigned to `retException`. If we change that now, it might be a
+                // breaking change. If not for v1, consider for #v2.
             }
+
+            return retException;
         }
 
-        public static async Task<T> GetObjectFromAmqpMessageAsync<T>(AmqpMessage amqpMessage)
+        public static ErrorContext GetErrorContextFromException(AmqpException exception)
         {
-            using var reader = new StreamReader(amqpMessage.BodyStream, Encoding.UTF8);
-            string jsonString = await reader.ReadToEndAsync().ConfigureAwait(false);
-            return JsonConvert.DeserializeObject<T>(jsonString);
+            Error error = exception.Error;
+            AmqpSymbol amqpSymbol = error.Condition;
+            string message = error.ToString();
+            if (Equals(AmqpErrorCode.ConnectionForced, amqpSymbol)
```

Comment:
Anything wrong with using `ampSymbol == AmqpErrorCode.ConnectionForced` instead of the `Equals` overload?

Reply:
That's simpler, yeah

Reply:
Actually, these error codes are Symbol instances, not enums, so we can't use "==" for comparison. We can do ```AmqpErrorCode.ConnectionForced.Equals(amqpSymbol)``` though